File: wasm/freeform_wasm.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <emscripten.h>

// Define implementation guard before including header
#define FF_FREEFORM_IMPL_
#include "../extern/free_form/freeform.h"

// Global sketch instance (managed by JavaScript)
static ff_Sketch* global_sketch = NULL;

// Initialize sketch
EMSCRIPTEN_KEEPALIVE
void* wasm_sketch_create(int p_cap, int e_cap, int c_cap) {
    ff_Sketch* skt = (ff_Sketch*)malloc(sizeof(ff_Sketch));
    if (skt) {
        ffSketch_Init(skt, p_cap, e_cap, c_cap);
    }
    return skt;
}

// Free sketch
EMSCRIPTEN_KEEPALIVE
void wasm_sketch_free(void* skt_ptr) {
    if (skt_ptr) {
        ff_Sketch* skt = (ff_Sketch*)skt_ptr;
        ffSketch_Free(skt);
        free(skt);
    }
}

// Solve sketch
EMSCRIPTEN_KEEPALIVE
int wasm_sketch_solve(void* skt_ptr, double tolerance, int max_steps) {
    if (!skt_ptr) return 0;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    return ffSketch_Solve(skt, tolerance, max_steps) ? 1 : 0;
}

// Add parameter
EMSCRIPTEN_KEEPALIVE
int wasm_add_parameter(void* skt_ptr, double value) {
    if (!skt_ptr) return -1;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    ff_ParameterDef def = {value};
    ff_ParamHandle h = ffSketch_AddParameter(skt, def);
    return h.idx;
}

// Get parameter value
EMSCRIPTEN_KEEPALIVE
double wasm_get_parameter(void* skt_ptr, int param_idx) {
    if (!skt_ptr) return 0.0;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    if (param_idx >= skt->params.cap || !skt->params.slots[param_idx].alive) {
        return 0.0;
    }
    return skt->params.slots[param_idx].payload.def.v;
}

// Set parameter value
EMSCRIPTEN_KEEPALIVE
void wasm_set_parameter(void* skt_ptr, int param_idx, double value) {
    if (!skt_ptr) return;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    if (param_idx >= skt->params.cap || !skt->params.slots[param_idx].alive) {
        return;
    }
    skt->params.slots[param_idx].payload.def.v = value;
}

// Add point entity
EMSCRIPTEN_KEEPALIVE
int wasm_add_point(void* skt_ptr, double x, double y) {
    if (!skt_ptr) return -1;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;

    ff_ParamHandle x_param = ffSketch_AddParameter(skt, (ff_ParameterDef){x});
    ff_ParamHandle y_param = ffSketch_AddParameter(skt, (ff_ParameterDef){y});

    ff_EntityDef pointDef = ff_EntityDef_DEFAULT(FF_POINT);
    pointDef.data.point.x = x_param;
    pointDef.data.point.y = y_param;

    ff_EntityHandle h = ffSketch_AddEntity(skt, pointDef);
    return h.idx;
}

// Add line entity
EMSCRIPTEN_KEEPALIVE
int wasm_add_line(void* skt_ptr, double x1, double y1, double x2, double y2) {
    if (!skt_ptr) return -1;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;

    // Create first point
    ff_ParamHandle x1_param = ffSketch_AddParameter(skt, (ff_ParameterDef){x1});
    ff_ParamHandle y1_param = ffSketch_AddParameter(skt, (ff_ParameterDef){y1});
    ff_EntityDef point1Def = ff_EntityDef_DEFAULT(FF_POINT);
    point1Def.data.point.x = x1_param;
    point1Def.data.point.y = y1_param;
    ff_EntityHandle p1 = ffSketch_AddEntity(skt, point1Def);

    // Create second point
    ff_ParamHandle x2_param = ffSketch_AddParameter(skt, (ff_ParameterDef){x2});
    ff_ParamHandle y2_param = ffSketch_AddParameter(skt, (ff_ParameterDef){y2});
    ff_EntityDef point2Def = ff_EntityDef_DEFAULT(FF_POINT);
    point2Def.data.point.x = x2_param;
    point2Def.data.point.y = y2_param;
    ff_EntityHandle p2 = ffSketch_AddEntity(skt, point2Def);

    // Create line
    ff_EntityDef lineDef = ff_EntityDef_DEFAULT(FF_LINE);
    lineDef.data.line.p1 = p1;
    lineDef.data.line.p2 = p2;

    ff_EntityHandle h = ffSketch_AddEntity(skt, lineDef);
    return h.idx;
}

// Add circle entity
EMSCRIPTEN_KEEPALIVE
int wasm_add_circle(void* skt_ptr, double x, double y, double radius) {
    if (!skt_ptr) return -1;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;

    // Create center point
    ff_ParamHandle x_param = ffSketch_AddParameter(skt, (ff_ParameterDef){x});
    ff_ParamHandle y_param = ffSketch_AddParameter(skt, (ff_ParameterDef){y});
    ff_EntityDef centerDef = ff_EntityDef_DEFAULT(FF_POINT);
    centerDef.data.point.x = x_param;
    centerDef.data.point.y = y_param;
    ff_EntityHandle center = ffSketch_AddEntity(skt, centerDef);

    // Create radius parameter
    ff_ParamHandle r_param = ffSketch_AddParameter(skt, (ff_ParameterDef){radius});

    // Create circle
    ff_EntityDef circleDef = ff_EntityDef_DEFAULT(FF_CIRCLE);
    circleDef.data.circle.c = center;
    circleDef.data.circle.r = r_param;

    ff_EntityHandle h = ffSketch_AddEntity(skt, circleDef);
    return h.idx;
}

// Get all parameters (returns count, fills arrays)
EMSCRIPTEN_KEEPALIVE
int wasm_get_all_parameters(void* skt_ptr, int* ids_out, double* values_out, int max_count) {
    if (!skt_ptr) return 0;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;

    int count = 0;
    for (int i = 0; i < skt->params.cap && count < max_count; i++) {
        if (skt->params.slots[i].alive) {
            ids_out[count] = i;
            values_out[count] = skt->params.slots[i].payload.def.v;
            count++;
        }
    }
    return count;
}

// Get all entities (returns count, fills arrays)
EMSCRIPTEN_KEEPALIVE
int wasm_get_all_entities(void* skt_ptr, int* ids_out, int* types_out, int max_count) {
    if (!skt_ptr) return 0;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;

    int count = 0;
    for (int i = 0; i < skt->entities.cap && count < max_count; i++) {
        if (skt->entities.slots[i].alive) {
            ids_out[count] = i;
            types_out[count] = skt->entities.slots[i].payload.def.type;
            count++;
        }
    }
    return count;
}

// Get entity data for a specific entity
EMSCRIPTEN_KEEPALIVE
int wasm_get_entity_type(void* skt_ptr, int entity_idx) {
    if (!skt_ptr) return -1;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    if (entity_idx >= skt->entities.cap || !skt->entities.slots[entity_idx].alive) {
        return -1;
    }
    return skt->entities.slots[entity_idx].payload.def.type;
}

// Get point entity data
EMSCRIPTEN_KEEPALIVE
void wasm_get_point_data(void* skt_ptr, int entity_idx, int* x_param_out, int* y_param_out) {
    if (!skt_ptr) return;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    if (entity_idx >= skt->entities.cap || !skt->entities.slots[entity_idx].alive) {
        return;
    }
    ff_Entity* ent = &skt->entities.slots[entity_idx].payload;
    if (ent->def.type == FF_POINT) {
        *x_param_out = ent->def.data.point.x.idx;
        *y_param_out = ent->def.data.point.y.idx;
    }
}

// Get line entity data
EMSCRIPTEN_KEEPALIVE
void wasm_get_line_data(void* skt_ptr, int entity_idx, int* p1_out, int* p2_out) {
    if (!skt_ptr) return;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    if (entity_idx >= skt->entities.cap || !skt->entities.slots[entity_idx].alive) {
        return;
    }
    ff_Entity* ent = &skt->entities.slots[entity_idx].payload;
    if (ent->def.type == FF_LINE) {
        *p1_out = ent->def.data.line.p1.idx;
        *p2_out = ent->def.data.line.p2.idx;
    }
}

// Get circle entity data
EMSCRIPTEN_KEEPALIVE
void wasm_get_circle_data(void* skt_ptr, int entity_idx, int* center_out, int* radius_param_out) {
    if (!skt_ptr) return;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    if (entity_idx >= skt->entities.cap || !skt->entities.slots[entity_idx].alive) {
        return;
    }
    ff_Entity* ent = &skt->entities.slots[entity_idx].payload;
    if (ent->def.type == FF_CIRCLE) {
        *center_out = ent->def.data.circle.c.idx;
        *radius_param_out = ent->def.data.circle.r.idx;
    }
}

// Get parameter count
EMSCRIPTEN_KEEPALIVE
int wasm_get_param_count(void* skt_ptr) {
    if (!skt_ptr) return 0;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    return skt->params.alive_count;
}

// Get entity count
EMSCRIPTEN_KEEPALIVE
int wasm_get_entity_count(void* skt_ptr) {
    if (!skt_ptr) return 0;
    ff_Sketch* skt = (ff_Sketch*)skt_ptr;
    return skt->entities.alive_count;
}

File: front/src/freeform-wasm.ts
// TypeScript wrapper for FreeForm WASM module

export interface FreeFormModule extends EmscriptenModule {
  _wasm_sketch_create(p_cap: number, e_cap: number, c_cap: number): number
  _wasm_sketch_free(skt_ptr: number): void
  _wasm_sketch_solve(skt_ptr: number, tolerance: number, max_steps: number): number

  _wasm_add_parameter(skt_ptr: number, value: number): number
  _wasm_get_parameter(skt_ptr: number, param_idx: number): number
  _wasm_set_parameter(skt_ptr: number, param_idx: number, value: number): void

  _wasm_add_point(skt_ptr: number, x: number, y: number): number
  _wasm_add_line(skt_ptr: number, x1: number, y1: number, x2: number, y2: number): number
  _wasm_add_circle(skt_ptr: number, x: number, y: number, radius: number): number

  _wasm_get_param_count(skt_ptr: number): number
  _wasm_get_entity_count(skt_ptr: number): number

  _wasm_get_all_parameters(skt_ptr: number, ids_out: number, values_out: number, max_count: number): number
  _wasm_get_all_entities(skt_ptr: number, ids_out: number, types_out: number, max_count: number): number

  _wasm_get_entity_type(skt_ptr: number, entity_idx: number): number
  _wasm_get_point_data(skt_ptr: number, entity_idx: number, x_param_out: number, y_param_out: number): void
  _wasm_get_line_data(skt_ptr: number, entity_idx: number, p1_out: number, p2_out: number): void
  _wasm_get_circle_data(skt_ptr: number, entity_idx: number, center_out: number, radius_param_out: number): void
}

declare function createFreeFormModule(): Promise<FreeFormModule>

export enum EntityType {
  POINT = 0,
  LINE = 1,
  CIRCLE = 2,
  ARC = 3
}

export interface Parameter {
  id: number
  value: number
}

export interface Entity {
  id: number
  type: EntityType
  data?: any
}

export class FreeFormSketch {
  private module: FreeFormModule | null = null
  private sketchPtr: number = 0

  async initialize(p_cap = 1024, e_cap = 256, c_cap = 128) {
    // @ts-ignore - Import WASM module
    const createModule = (await import('/freeform.js')).default
    this.module = await createModule()
    this.sketchPtr = this.module._wasm_sketch_create(p_cap, e_cap, c_cap)
  }

  destroy() {
    if (this.module && this.sketchPtr) {
      this.module._wasm_sketch_free(this.sketchPtr)
      this.sketchPtr = 0
    }
  }

  solve(tolerance = 0.01, maxSteps = 8): boolean {
    if (!this.module || !this.sketchPtr) return false
    return this.module._wasm_sketch_solve(this.sketchPtr, tolerance, maxSteps) === 1
  }

  addParameter(value: number): number {
    if (!this.module || !this.sketchPtr) return -1
    return this.module._wasm_add_parameter(this.sketchPtr, value)
  }

  getParameter(paramId: number): number {
    if (!this.module || !this.sketchPtr) return 0
    return this.module._wasm_get_parameter(this.sketchPtr, paramId)
  }

  setParameter(paramId: number, value: number) {
    if (!this.module || !this.sketchPtr) return
    this.module._wasm_set_parameter(this.sketchPtr, paramId, value)
  }

  addPoint(x: number, y: number): number {
    if (!this.module || !this.sketchPtr) return -1
    return this.module._wasm_add_point(this.sketchPtr, x, y)
  }

  addLine(x1: number, y1: number, x2: number, y2: number): number {
    if (!this.module || !this.sketchPtr) return -1
    return this.module._wasm_add_line(this.sketchPtr, x1, y1, x2, y2)
  }

  addCircle(x: number, y: number, radius: number): number {
    if (!this.module || !this.sketchPtr) return -1
    return this.module._wasm_add_circle(this.sketchPtr, x, y, radius)
  }

  getAllParameters(): Parameter[] {
    if (!this.module || !this.sketchPtr) return []

    const count = this.module._wasm_get_param_count(this.sketchPtr)
    if (count === 0) return []

    const idsPtr = this.module._malloc(count * 4)
    const valuesPtr = this.module._malloc(count * 8)

    const actualCount = this.module._wasm_get_all_parameters(this.sketchPtr, idsPtr, valuesPtr, count)

    const ids = new Int32Array(this.module.HEAP32.buffer, idsPtr, actualCount)
    const values = new Float64Array(this.module.HEAPF64.buffer, valuesPtr, actualCount)

    const parameters: Parameter[] = []
    for (let i = 0; i < actualCount; i++) {
      parameters.push({ id: ids[i], value: values[i] })
    }

    this.module._free(idsPtr)
    this.module._free(valuesPtr)

    return parameters
  }

  getAllEntities(): Entity[] {
    if (!this.module || !this.sketchPtr) return []

    const count = this.module._wasm_get_entity_count(this.sketchPtr)
    if (count === 0) return []

    const idsPtr = this.module._malloc(count * 4)
    const typesPtr = this.module._malloc(count * 4)

    const actualCount = this.module._wasm_get_all_entities(this.sketchPtr, idsPtr, typesPtr, count)

    const ids = new Int32Array(this.module.HEAP32.buffer, idsPtr, actualCount)
    const types = new Int32Array(this.module.HEAP32.buffer, typesPtr, actualCount)

    const entities: Entity[] = []
    for (let i = 0; i < actualCount; i++) {
      const entity: Entity = {
        id: ids[i],
        type: types[i]
      }

      // Get type-specific data
      const dataPtr = this.module._malloc(16) // 4 ints max
      const dataView = new Int32Array(this.module.HEAP32.buffer, dataPtr, 4)

      switch (entity.type) {
        case EntityType.POINT:
          this.module._wasm_get_point_data(this.sketchPtr, entity.id, dataPtr, dataPtr + 4)
          entity.data = {
            x_param: dataView[0],
            y_param: dataView[1]
          }
          break
        case EntityType.LINE:
          this.module._wasm_get_line_data(this.sketchPtr, entity.id, dataPtr, dataPtr + 4)
          entity.data = {
            p1: dataView[0],
            p2: dataView[1]
          }
          break
        case EntityType.CIRCLE:
          this.module._wasm_get_circle_data(this.sketchPtr, entity.id, dataPtr, dataPtr + 4)
          entity.data = {
            center: dataView[0],
            radius_param: dataView[1]
          }
          break
      }

      this.module._free(dataPtr)
      entities.push(entity)
    }

    this.module._free(idsPtr)
    this.module._free(typesPtr)

    return entities
  }
}

// Create singleton instance
let sketchInstance: FreeFormSketch | null = null

export async function getSketch(): Promise<FreeFormSketch> {
  if (!sketchInstance) {
    sketchInstance = new FreeFormSketch()
    await sketchInstance.initialize()
  }
  return sketchInstance
}